runtime/libc/wasi/include/wasi_backing.h

```diff
@@ -0,0 +1,1320 @@
+        goto done;
+
+    /* Write Result */
+    wasi_serdes_write_size_t(CURRENT_MEMORY_BASE, nbytes_retptr, nread);
+
+done:
+    return (uint32_t)rc;
+}
+
+/**
+ * Return a description of the given preopened file descriptor.
+ *
+ * @param fd
+ * @param prestat_retptr The buffer where the description is stored.
+ * @return status code
+ */
+uint32_t wasi_snapshot_preview1_fd_prestat_get(__wasi_fd_t fd, __wasi_size_t prestat_retptr) {
+#ifdef LOG_WASI
+    fprintf(stderr, "wasi_snapshot_preview1_fd_prestat_get(%u,%u)\n", fd, prestat_retptr);
+#endif
+
+    /* Check Bounds */
+    wasi_serdes_check_bounds(prestat_retptr, CURRENT_MEMORY_SIZE, WASI_SERDES_SIZE_prestat_t);
+
+    /* Execute WASI call */
+    __wasi_prestat_t prestat;
+    __wasi_errno_t   rc = __wasi_fd_prestat_get(CURRENT_WASI_CONTEXT, fd, &prestat);
+    if (rc != __WASI_ERRNO_SUCCESS)
+        goto done;
+
+    /* Write result to linear memory */
+    wasi_serdes_write_prestat_t(CURRENT_MEMORY_BASE, prestat_retptr, &prestat);
+
+done:
+    return (uint32_t)rc;
+}
+
+/**
+ * Return a description of the given preopened file descriptor.
+ *
+ * @param fd
+ * @param path_retptr A buffer into which to write the preopened directory name.
+ * @param path_len The length of the buffer at path_retptr
+ * @return status code
+ */
+uint32_t wasi_snapshot_preview1_fd_prestat_dir_name(__wasi_fd_t fd, __wasi_size_t path_retptr, __wasi_size_t path_len) {
+#ifdef LOG_WASI
+    fprintf(stderr, "wasi_snapshot_preview1_fd_prestat_dir_name(%u,%u,%u)\n", fd, path_retptr, path_len);
+#endif
+
+    /* Check Bounds */
+    wasi_serdes_check_bounds(path_retptr, CURRENT_MEMORY_SIZE, path_len);
+
+    /* Execute WASI call, writing results directly to linear memory */
+    return (uint32_t)__wasi_fd_prestat_dir_name(CURRENT_WASI_CONTEXT, fd, (char*)&CURRENT_MEMORY_BASE[path_retptr],
+                                                path_len);
+}
+
+/**
+ * Write to a file descriptor without updating the descriptor's offset
+ *
+ * @param fd
+ * @param iovs_baseptr List of scatter/gather vectors from which to retrieve data.
+ * @param iovs_len The length of the array pointed to by `iovs`.
+ * @param offset The offset within the file at which to write.
+ * @param nwritten_retptr The number of bytes written.
+ * @return status code
+ *
+ */
+uint32_t wasi_snapshot_preview1_fd_pwrite(__wasi_fd_t fd, __wasi_size_t iovs_baseptr, __wasi_size_t iovs_len,
+                                          __wasi_filesize_t offset, __wasi_size_t nwritten_retptr) {
+#ifdef LOG_WASI
+    fprintf(stderr, "wasi_snapshot_preview1_fd_pwrite(%u,%u,%u,%lu,%u)\n", fd, iovs_baseptr, iovs_len, offset,
+            nwritten_retptr);
+#endif
+
+    /* Check Bounds */
+    wasi_serdes_check_bounds(iovs_baseptr, CURRENT_MEMORY_SIZE, iovs_len * WASI_SERDES_SIZE_ciovec_t);
+    wasi_serdes_check_bounds(nwritten_retptr, CURRENT_MEMORY_SIZE, WASI_SERDES_SIZE_size_t);
+
+    /* Copy iovs into memory */
+    __wasi_ciovec_t iovs[iovs_len];
```

Comment:
I'd have to dig back in to be sure, but I believe that the iovecs need to be copied because the serialization layer does "fixups" to replace memory offsets with host addresses. Perhaps some of these intermediate copied can be removed, but it'd require some concentrated effort. I've added #53 as a follow on effort for this. I understand this is critical, but I still think this should be deferred until after baseline WASI support is merged into aWsm and SLEdge.